File: handle/router_controller_reflection_route_model_binding.py
import inspect

class UserController:
    def get_user_by_id(self, id: int):
        return f"Get user by ID: {id}"

    def get_user_by_name(self, name: str):
        return f"Get user by name: {name}"

def analyze_controller_method(controller_class, method_name):
    try:
        method = getattr(controller_class(), method_name)
        parameters = inspect.signature(method).parameters
        parameter_types = {name: param.annotation for name, param in parameters.items() if param.annotation != inspect.Parameter.empty}
        return parameter_types
    except AttributeError:
        return None

# Phân tích phương thức get_user_by_id trong UserController
user_controller_params = analyze_controller_method(UserController, 'get_user_by_id')
if user_controller_params is not None:
    print("get_user_by_id parameter types:")
    print(user_controller_params)
else:
    print("Method get_user_by_id not found in UserController")

# Phân tích phương thức get_user_by_name trong UserController
user_controller_params = analyze_controller_method(UserController, 'get_user_by_name')
if user_controller_params is not None:
    print("\nget_user_by_name parameter types:")
    print(user_controller_params)
else:
    print("Method get_user_by_name not found in UserController")

# Phân tích phương thức không tồn tại trong UserController
non_existent_method_params = analyze_controller_method(UserController, 'non_existent_method')
if non_existent_method_params is not None:
    print("\nnon_existent_method parameter types:")
    print(non_existent_method_params)
else:
    print("Method non_existent_method not found in UserController")

File: handle/router_controller_reflection_route_model_binding.php
<?php

class UserController
{
    public function getUserById(int $id)
    {
        return "Get user by ID: $id";
    }

    public function getUserByName(string $name)
    {
        return "Get user by name: $name";
    }
}

function analyzeControllerMethod($controllerClassName, $methodName)
{
    $reflectionClass = new ReflectionClass($controllerClassName);
    $reflectionMethod = $reflectionClass->getMethod($methodName);

    $parameters = $reflectionMethod->getParameters();
    $parameterTypes = [];

    foreach ($parameters as $parameter) {
        $parameterName = $parameter->getName();
        $parameterType = $parameter->getType();

        if ($parameterType) {
            $parameterTypes[$parameterName] = $parameterType->getName();
        } else {
            $parameterTypes[$parameterName] = 'mixed'; // Nếu không có kiểu dữ liệu được định nghĩa, mặc định là 'mixed'
        }
    }

    return $parameterTypes;
}

// Phân tích phương thức getUserById trong UserController
$userControllerParams = analyzeControllerMethod('UserController', 'getUserById');
echo "getUserById parameter types: \n";
print_r($userControllerParams);

// Phân tích phương thức getUserByName trong UserController
$userControllerParams = analyzeControllerMethod('UserController', 'getUserByName');
echo "\ngetUserByName parameter types: \n";
print_r($userControllerParams);


//call fuction via name , and name agrument prams
function greet($message, $name) {
    echo "Hello $name, $message!";
}

$args = [
    'name' => 'John',
    'message' => 'how are you?'
];

call_user_func_array('greet', $args);
//c2

<?php
function makeyogurt($container: int)
{
    return  $container
}
function makeyogurt($container: str)
{
    return  $container
}

echo makeyogurt(style: "natural");
?>
